### PR TITLE
Fix a race in kube-proxy causing runaways

### PR DESCRIPTION
**Pull request by thockin**
It was an ABA problem where the proxy loop might see its own service as
"existing" when it had been destroyed and recreated (as in an update).

To prove this I added a counter of running ProxyLoop goroutines and check that
in tests.  If I undo my main change, the tests fail.  This makes the
proxier_test significantly slower (3 seconds vs 0.5 seconds).  Sorry.

Mitigation for #3943

@dchen1107  @phemmer

### CODE REVIEW

**Comment by thockin** [comment]
ping - I think this is a pretty significant bug and it's an easy PR :)

File: pkg/proxy/proxier.go [hunk 1]
```diff
@@ -300,12 +301,13 @@ func newProxySocket(protocol api.Protocol, ip net.IP, port int) (proxySocket, er
...
-	loadBalancer LoadBalancer
-	mu           sync.Mutex // protects serviceMap
-	serviceMap   map[string]*serviceInfo
-	listenIP     net.IP
-	iptables     iptables.Interface
-	hostIP       net.IP
+	loadBalancer  LoadBalancer
+	mu            sync.Mutex // protects serviceMap
+	serviceMap    map[string]*serviceInfo
+	numProxyLoops int32 // use atomic ops to access this
```
Comment (brendandburns) [inline pkg/proxy/proxier.go]: Can you note that this is only used for testing (and in the waitFor... function too)

**Comment by brendandburns** [comment]
LGTM.  one small nit.  Feel free to self-merge once its addressed.

File: pkg/proxy/proxier.go (hunk 1 above)
Comment (thockin) [inline pkg/proxy/proxier.go]: Will do

**Comment by thockin** [comment]
Pushed with added comment.  Merging now.

**Pull request merged**